### ISSUE TITLE
No gradient defined for op: Select

### ISSUE THREAD

**Issue by gundamMC** [migueldeicaza/TensorFlowSharp#264]
It's a similar issue to #187 but it is the operation "Select" instead of "Pow". 
The error
`No gradient defined for op: Select. Please see 
https://www.tensorflow.org/code/tensorflow/cc/gradients/README.md 
for instructions on how to add C++ gradients.`
shows up when using AddGradients with TensorFlowSharp v1.7.0.

Here's my code: https://github.com/gundamMC/Neural-Network-GUI-Demo/blob/master/NeuralNetwork/Network.cs
Everything is pretty much covered in this single file.

I don't know if this matters but there is already an issue in the Tensorflow repo (tensorflow/tensorflow#14845) and a pr which states that it's fixed (tensorflow/tensorflow#14862).

**Comment by wmeints** [comment]
Experiencing the exact same issue. Would love to use a workaround, but not sure how to do that.

**Comment by bendiy** [comment]
This error still exists in the current `master`. However, tensorflow merged the fix in December of 2017, so I'm not sure why there's still an error in TensorFlowSharp.

@gundamMC were you able to get it to work? You removed this from your README.md. When I tried your repo, I still get the same error.
```diff
-*still in development, meaning that it will not work*
```
- https://github.com/gundamMC/Neural-Network-GUI-Demo/commit/885ffd700d94720a93abd7eb97e49bcf33a34464

I combined the @gundamMC example with the one from #187 to create a self contained example that reproduces the error with `Select`:

```c#
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using TensorFlow;

namespace LinearFit
{
    class Program
    {
        public static void Main(string[] args)
        {
            // Test for: https://github.com/migueldeicaza/TensorFlowSharp/issues/264

            TFOutput Predict(TFOutput data, TFOutput[] weights, TFOutput[] biases, TFGraph graph)
            {
                TFOutput LayerOut = data;

                for (int i = 0; i < weights.Length; i++)
                {
                    LayerOut = graph.Add(graph.MatMul(LayerOut, weights[i]), biases[i]);
                }

                return LayerOut;
            }

            var xList = new List<double[]>();
            var yList = new List<double[]>();
            var ran = new Random();
            for (var i = 0; i < 4; i++)
            {
                var foo = new double[100];

                for (var j = 0; j < 100; j++)
                {
                    var num = (double)ran.NextDouble();
                    var noise = (double)ran.NextDouble();
                    foo[j] = num * 3 + 4 + noise;
                }

                xList.Add(foo);
            }

            for (var i = 0; i < 4; i++)
            {
                var foo = new double[6];

                for (var j = 0; j < 6; j++)
                {
                    var num = (double)ran.NextDouble();
                    var noise = (double)ran.NextDouble();
                    foo[j] = num * 3 + 4 + noise;
                }

                yList.Add(foo);
            }

            var xData = xList.ToArray();
            var yData = yList.ToArray();
            var learning_rate = 0.01;

            var g = new TFGraph();

            int[] NodeSize = new int[] { 7 };
            int InputSize = xData[0].Length;
            int OutputSize = yData[0].Length;

            var x = g.Placeholder(TFDataType.Double, new TFShape(new long[] { -1, InputSize }));
            var y = g.Placeholder(TFDataType.Double, new TFShape(new long[] { -1, OutputSize }));

            TFOutput[] W = new TFOutput[NodeSize.Length + 1];
            TFOutput[] b = new TFOutput[NodeSize.Length + 1];

            int prevSize = InputSize;
            for (int i = 0; i < NodeSize.Length; i++)
            {
                W[i] = g.VariableV2(new TFShape(new long[] { prevSize, NodeSize[i] }), TFDataType.Double, operName: "weight_" + i);
                b[i] = g.VariableV2(new TFShape(new long[] { NodeSize[i] }), TFDataType.Double, operName: "bias_" + i);
                prevSize = NodeSize[i];
            }

            W[NodeSize.Length] = g.VariableV2(new TFShape(new long[] { prevSize, OutputSize }), TFDataType.Double, operName: "weight_out");
            b[NodeSize.Length] = g.VariableV2(new TFShape(new long[] { OutputSize }), TFDataType.Double, operName: "bias_out");

            TFOutput output = Predict(x, W, b, g);
            
            TFOutput loss = g.ReduceMean(g.SigmoidCrossEntropyWithLogits(y, output));

            TFOutput[] parameters = new TFOutput[W.Length + b.Length];
            W.CopyTo(parameters, 0);
            b.CopyTo(parameters, W.Length);

            TFOutput[] grad = g.AddGradients(new TFOutput[] { loss }, parameters);

            TFOperation[] optimize = new TFOperation[parameters.Length];

            for (int i = 0; i < parameters.Length; i++)
            {
                optimize[i] = g.AssignSub(parameters[i], g.Mul(grad[i], g.Const(learning_rate))).Operation;
            }

            var sess = new TFSession(g);

            TFOperation[] InitParams = new TFOperation[parameters.Length];

            for (int i = 0; i < parameters.Length; i++)
            {
                InitParams[i] = g.Assign(parameters[i], g.RandomNormal(g.GetTensorShape(parameters[i]))).Operation;
            }
            
            for (var i = 0; i < 1000; i++)
            {
                TFTensor result = sess.GetRunner()
                    .AddInput(x, xData)
                    .AddInput(y, yData)
                    .AddTarget(optimize)
                    .Fetch(loss)
                    .Run();

                Console.WriteLine("Epoch - " + i + " loss: {0} ", result.GetValue());
            }
        }
    }
}
```

**Comment by gundamMC** [comment]
@bendiy I was going to try something to fix it, but I never got around to it. Have you tried using the new SGD optimizer introduced in e7e8f9f?

They have a sample code [here](https://github.com/migueldeicaza/TensorFlowSharp/blob/e7e8f9f715f56ff4bb0cab3a5c02dbb3d5ffab9f/SampleTest/SampleTest.cs#L329).

**Comment by bendiy** [comment]
@gundamMC Yes, the new SGD works without error.
